pkg/server/handler.go

```diff
@@ -253,6 +256,83 @@ func WithRequestIdentity(handler http.Handler) http.Handler {
 	})
 }
 
+type privilege int
+
+const (
+	unprivileged privilege = iota
+	priviledged
+	superPrivileged
+)
+
+var (
+	// specialGroups specify groups with special meaning kcp. Lower privilege (= higher number)
+	// cannot impersonate higher privilege levels.
+	specialGroups = map[string]privilege{
+		authorizationbootstrap.SystemMastersGroup:  superPrivileged,
+		authorizationbootstrap.SystemKcpAdminGroup: priviledged,
+	}
+)
+
+// WithImpersonationGatekeeper checks the request for impersonations and validates them,
+// if they are valid. If they are not, will return a 403.
+// We check for impersonation in the request headers, early to avoid it being propagated to
+// the backend services.
+func WithImpersonationGatekeeper(handler http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		// Impersonation check is only done when impersonation is requested.
+		// And impersonations is only allowed for the users, who have metadata in the ctx.
+		// Else just pass the request.
+		impersonationUser := req.Header.Get(authenticationv1.ImpersonateUserHeader)
+		if len(impersonationUser) == 0 {
+			handler.ServeHTTP(w, req)
+			return
+		}
+		requester, exists := request.UserFrom(req.Context())
+		if !exists {
+			responsewriters.ErrorNegotiated(
+				apierrors.NewForbidden(schema.GroupResource{}, "", fmt.Errorf("impersonation is invalid for the requestor")),
+				errorCodecs, schema.GroupVersion{}, w, req)
+			return
+		}
+		if validImpersonation(requester.GetGroups(), req.Header[authenticationv1.ImpersonateGroupHeader]) {
+			handler.ServeHTTP(w, req)
+			return
+		}
+
+		responsewriters.ErrorNegotiated(
+			apierrors.NewForbidden(schema.GroupResource{}, "", fmt.Errorf("impersonation is not allowed for the requestor")),
+			errorCodecs, schema.GroupVersion{}, w, req)
+	})
+}
+
+// maxUserPrivilege returns the highest privilege level found among the user's groups.
+func maxUserPrivilege(userGroups []string) privilege {
+	max := unprivileged
+	for _, g := range userGroups {
+		if p, found := specialGroups[g]; found && p > max {
+			max = p
+		}
+	}
+	return max
+}
+
+// validImpersonation checks if a user can impersonate all requested groups.
+func validImpersonation(userGroups, requestedGroups []string) bool {
+	userMax := maxUserPrivilege(userGroups)
+
+	for _, g := range requestedGroups {
+		p := unprivileged
+		if sp, found := specialGroups[g]; found {
```

Comment:
no need for this. Just do `p := specialGroups[g]`. Zero value is `unprivileged`.